### PR TITLE
Initial add of hearthstone

### PR DESCRIPTION
**Pull request by SethKitchen**
builds and runs -- currently stuck because

[cuda-24-0.local:08642] 2 more processes have sent help message help-mpi-btl-openib.txt / no active ports found
[cuda-24-0.local:08642] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
slurmstepd: Job 1171978 exceeded memory limit (18202456 > 16384000), being killed
slurmstepd: *** JOB 1171978 ON cuda-24-0 CANCELLED AT 2018-10-16T13:24:27 ***
slurmstepd: Exceeded step memory limit at some point.

Why using so much memory?

### CODE REVIEW

**Comment by benbellheron** [comment]
Seth, thanks for this PR! The Hearthstone environment will be a great addition to Adept. I'll look over this PR tomorrow and hopefully get it merged in.

IMPALA sends experiences (state, action, log_prob of action, rewards, network internals [lstm state], and terminals) from each worker to the host. Depending on the number of environments and state space this can be a lot of data. Towered is much better in terms of total communication overhead (it's just gradients) and can be just as fast or faster, not to mention requires much less VRAM. It really depends on the setup of your cluster. IMPALA is good for lots of CPU based workers and one GPU machine. Towered is better for multiple similar machines that have GPUs.

File: idk [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# This script will uninstall all the versions of Python 3.x
```
[the remark this answers is not in the document]
Comment (SethKitchen) [reply idk]: agreed -- didn't mean to push this

File: adept/containers/impala.py [hunk 2]
```diff
@@ -253,13 +253,19 @@ def stop_mpi_workers(self):
     def get_parameter_shapes(self):
         shapes = [tuple(x.shape) for x in self.network.parameters()]
         if not self.use_local_buffers:
-            shapes.extend([tuple(x.shape) for x in self.network._all_buffers()])
+            if torch.__version__=='1.0.0a0+16b8075':
```
[the remark this answers is not in the document]
Comment (SethKitchen) [reply adept/containers/impala.py]: done

File: adept/containers/impala.py [hunk 3]
```diff
@@ -253,13 +253,19 @@ def stop_mpi_workers(self):
...
+                shapes.extend([tuple(x.shape) for x in self.network.buffers()])
+            else:
+                shapes.extend([tuple(x.shape) for x in self.network._all_buffers()])         
         return shapes
 
     def get_parameters_numpy(self):
         params = [p.detach().cpu().numpy() for p in self.network.parameters()]
         if not self.use_local_buffers:
-            params.extend([b.cpu().numpy() for b in self.network._all_buffers()])
+            if torch.__version__=='1.0.0a0+16b8075':
```
[the remark this answers is not in the document]
Comment (SethKitchen) [reply adept/containers/impala.py]: done

File: adept/containers/impala.py [hunk 4]
```diff
@@ -427,15 +433,21 @@ def receive(self):
     def set_parameters(self, parameters):
         local_params = list(self.network.parameters())
         if not self.use_local_buffers:
-            local_params.extend([b for b in self.network._all_buffers()])
+            if torch.__version__=='1.0.0a0+16b8075':
```
[the remark this answers is not in the document]
Comment (SethKitchen) [reply adept/containers/impala.py]: done

File: adept/containers/impala.py [hunk 5]
```diff
@@ -427,15 +433,21 @@ def receive(self):
...
+                local_params.extend([b for b in self.network._all_buffers()])
 
         for p, v in zip(local_params, parameters):
             p.data.copy_(v, non_blocking=True)
 
     def get_parameter_shapes(self):
         shapes = [tuple(x.shape) for x in self.network.parameters()]
         if not self.use_local_buffers:
-            shapes.extend([tuple(x.shape) for x in self.network._all_buffers()])
+            if torch.__version__=='1.0.0a0+16b8075':
```
[the remark this answers is not in the document]
Comment (SethKitchen) [reply adept/containers/impala.py]: done

File: adept/containers/towered.py [hunk 6]
```diff
@@ -48,7 +48,11 @@ def __init__(self, mpi_comm, num_grads_to_drop, network, make_optimizer, saver,
         self.variable_flattener = ArrayFlattener([tuple(x.shape) for x in self.network.parameters()])
 
         # vars for sending network buffers
-        self.mpi_buffer_sender = MPIArraySend(mpi_comm, [tuple(x.shape) for x in 
+        if torch.__version__=='1.0.0a0+16b8075':
```
[the remark this answers is not in the document]
Comment (SethKitchen) [reply adept/containers/towered.py]: done

File: adept/containers/towered.py [hunk 7]
```diff
@@ -169,7 +173,11 @@ def run(self, max_steps=float('inf'), initial_step_count=0):
     def _saver_thread(self):
         num_workers = self.comm.Get_size() - 1
         # setup for receiving buffers
-        buffer_shapes = [tuple(x.shape) for x in self.network._all_buffers()]
+        if torch.__version__=='1.0.0a0+16b8075':
```
[the remark this answers is not in the document]
Comment (SethKitchen) [reply adept/containers/towered.py]: done

File: adept/containers/towered.py [hunk 8]
```diff
@@ -202,9 +210,14 @@ def _saver_thread(self):
                     # can't divide here since numpy reduces from array to float on tensors with shape ()
                     all_buffer_params = [x for x in unflattened_buffer_params]
                     # set buffers
-                    for b, all_bp in zip(self.network._all_buffers(), all_buffer_params):
-                        # mean over all workers
-                        b.copy_(torch.from_numpy(all_bp)).div_(num_workers)
+                    if torch.__version__=='1.0.0a0+16b8075':
```
[the remark this answers is not in the document]
Comment (SethKitchen) [reply adept/containers/towered.py]: done

File: adept/containers/towered.py [hunk 9]
```diff
@@ -274,8 +287,13 @@ def __init__(
             send_shapes, recv_shapes, 0, max_parameter_skip, gradient_warning_time, variable_warning_time
         )
         self.mpi_buffer_request = self._create_mpi_buffer_request()
-        self.mpi_buffer_sender = MPIArraySend(self._mpi_comm, [tuple(x.shape) for x in
-                                                               self.network._all_buffers()])
+        if torch.__version__=='1.0.0a0+16b8075':
```
[the remark this answers is not in the document]
Comment (SethKitchen) [reply adept/containers/towered.py]: done

File: adept/containers/towered.py [hunk 10]
```diff
@@ -349,7 +367,11 @@ def submit(self):
             self.global_step = 0
         # host decides when it wants pytorch buffers
         if self.mpi_buffer_request.test()[0]:
-            buffer_list = [x.cpu().numpy() for x in self.network._all_buffers()]
+            if torch.__version__=='1.0.0a0+16b8075':
```
[the remark this answers is not in the document]
Comment (SethKitchen) [reply adept/containers/towered.py]: done

File: adept/environments/hearthstone.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    return _f
+
+
+def hearthstone_env():
+    env = gym.make('Hearthstone-v0')
+    env = AdeptGymEnv(env)
+    return env
+
+
+class AdeptGymEnv(BaseEnvironment):
```
[the remark this answers is not in the document]
Comment (SethKitchen) [reply adept/environments/hearthstone.py]: done
